pkg/validation/test.go

```diff
@@ -30,30 +30,39 @@ const (
 )
 
 // context contains the information from parent components.
-// All but `fieldRoot` can be nil if the validation being performed in
+// All but `field` can be nil if the validation being performed in
 // context-independent.
 type context struct {
-	fieldRoot  string
+	field      fieldPath
 	env        api.TestEnvironment
 	seen       sets.String
 	leasesSeen sets.String
 	releases   sets.String
 }
 
-func newContext(fieldRoot string, env api.TestEnvironment, releases sets.String) context {
-	return context{
-		fieldRoot:  fieldRoot,
+// newContext creates a top-level context.
+func newContext(field fieldPath, env api.TestEnvironment, releases sets.String) *context {
```

Comment:
This is a value type in my mind, but Go would not let me use `newContext(…).errorf` because of the pointer receiver.

I gave up trying to fight it and changed all return types to pointers.

Reply:
Hopefully things get reasonably optimized since it's private to the module.